### DISCUSSION TITLE
Trimming FFmpeg Compilation

### DISCUSSION THREAD

**Discussion by Tryanks** [asticode/go-astiav#50] [Q&A]
For various reasons, I started to explore more possibilities. After statically linking a Go program to FFmpeg, it ballooned to 120MB in size. Therefore, I wanted to use the rich customizable options of FFmpeg's `./configure` to disable the compilation of some components.

Based on the components currently making up the pkgconfig:

```
pkg-config --cflags libavfilter libavutil libavcodec libavformat libavdevice
```

I attempted to exclude libraries unrelated to these:

```
./configure \
    --enable-static \
    --disable-programs \
    --disable-doc \
    --disable-swresample \
    --disable-swscale \
    --disable-postproc \
    --disable-pthreads \
    --disable-w32threads \
    --disable-os2threads \
    --disable-network
```

However, I encountered a compilation error: `cannot find '-lswscale'`. So, I removed `--disable-swscale` from the compile script and tried again. This time, the error was: `cannot find '-lswresample'`.

These are all my attempts at trimming FFmpeg. Through these efforts, I've found it challenging to define the scope of trimming for the FFmpeg libraries.

With this context, I pose the question:
Can anyone help explore the minimal FFmpeg trimming that would still allow `go-astiav` to function properly?

**Comment by asticode** [comment]
Would you mind sharing the command(s) you used to link your Go program to ffmpeg? That could be an interesting section to add in the README 👍 

> This time, the error was: cannot find '-lswresample'

Did you try to remove `--disable-swresample` from the compile script and if yes, did it work afterwards?

> Can anyone help explore the minimal FFmpeg trimming that would still allow go-astiav to function properly?

I won't have enough time to spend on this time-consuming issue in the short term but I'm really interested in your conclusions. My guess is that even if you use a portion of `astiav`'s codebase, Go will still need to compile the C portion of the unused code 🤔 Therefore there won't have much libs you'll be able to trim. Also my guess is that you're pretty safe using `--disable-everything` which would disable all codecs, protocols, etc. and enable only the one you need although I don't know how much size you'll gain on the final binary.

**Reply by Tryanks** [reply]
I'd be more than happy to share the way I used to compiler ffmpeg. I'll circle back to this soon after clearing up my recent workload.
But, as what you see, it just run `./configure --args` then `make && sudo make install`.

> Did you try to remove `--disable-swresample` from the compile script and if yes, did it work afterwards?

I did indeed try removing `--disable-swresample` from the compile script again, and it successfully passed the compilation for `go-astiav`, though it didn't significantly change the size of the compiled package.

> Also my guess is that you're pretty safe using `--disable-everything` which would disable all codecs, protocols, etc. and enable only the one you need although I don't know how much size you'll gain on the final binary.

I find this conjecture quite intriguing 👍 and plan to test its feasibility within this week.

I understand the constraints of time and appreciate your interest. Since time limitation is also a major challenge I face, I will be sure to share my findings and conclusions once I've conducted the necessary experiments.

**Comment by Tryanks** [comment]
Update on September 13, 2024: After being occupied elsewhere for a while, I returned here to find that the main branch has caught up with FFmpeg 7.0. Consequently, I carried out a compilation test using the latest version.

The testing process used [examples/transcoding/main.go](https://github.com/asticode/go-astiav/blob/master/examples/transcoding/main.go)

Step 1: Used just './configure' for initial setup.
Compilation commands:
```bash
./configure --enable-static # FFmpeg
go build -ldflags '-extldflags "-static"' . # astiav Go
```
Compilation results:
```
# go clean -cache -testcache
# go build -ldflags '-extldflags "-static"' .
# du -sh main
60M     main
```

Step 2: Removed non-dependent components based on pkg-config flags currently marked in `astiav.go`.
Compilation commands:
```bash
./configure --enable-static --disable-programs --disable-doc --disable-swresample --disable-postproc --disable-pthreads --disable-w32threads --disable-os2threads --disable-network
```
Compilation results:
```
# go clean -cache -testcache
# go build -ldflags '-extldflags "-static"' .
/usr/bin/ld: /tmp/go-link-1022824679/000102.o: in function `_cgo_04fbb8f65a5f_C2func_getaddrinfo':
/tmp/go-build/cgo-gcc-prolog:60: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
# du -sh main
58M     main
```
Cool, there wasn’t a significant change in size. However, unlike before, **this package does not depend on `-lswresample` anymore**, which allows me to follow pkg-config more closely for trimming.

Step 3: Removed symbol table; based on the settings from step 2, added `-ldflags="-s -w"`
Compilation results:
```
# go clean -cache -testcache
# go build -ldflags '-extldflags "-static" -s -w' .
/usr/bin/ld: /tmp/go-link-1022824679/000102.o: in function `_cgo_04fbb8f65a5f_C2func_getaddrinfo':
/tmp/go-build/cgo-gcc-prolog:60: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
# du -sh main
25M     main
```
It is quite surprising that debugging information occupied such a large portion of the binary file. Currently, a 25MB binary file looks very compelling, which somewhat brings into question the necessity for further trimming.

Additionally, after reading [this comment](https://github.com/asticode/go-astiav/pull/70#discussion_r1744734121) about PR#70, I suspect that completely removing the `#cgo pkgconfig:` tag and leaving it to the user space to handle might help further trim unnecessary modules (e.g., cutting all filters for purely transcoding applications). I will continue exploring this approach.
